### PR TITLE
Add caching in EntityIdSingleton

### LINKED ISSUE #11166: Possibly CATASTROPHIC failure - Cannot create 0x00000000000000000000000000000000005b7220 with number 5992993 (5992992) pending observed on testnet
### Description

The following error with the exception was observed twice for the past 12 hours in Grafana logs on testnet 
`2025-05-13 13:51:34.434	
2025-05-13T10:51:34.433Z ERROR tomcat-handler-417126 c.h.n.a.w.h.DispatchProcessor Possibly CATASTROPHIC failure - exception thrown while handling dispatch java.lang.IllegalStateException: Cannot create 0x00000000000000000000000000000000005b7220 with number 5992993 (5992992) pending
	at com.hedera.node.app.service.contract.impl.state.ProxyWorldUpdater.getValidatedCreationNumber(ProxyWorldUpdater.java:493)
	at com.hedera.node.app.service.contract.impl.state.ProxyWorldUpdater.createAccount(ProxyWorldUpdater.java:343)
	at org.hyperledger.besu.evm.worldstate.WorldUpdater.createAccount(WorldUpdater.java:64)
	at org.hyperledger.besu.evm.worldstate.WorldUpdater.getOrCreate(WorldUpdater.java:77)
	at com.hedera.node.app.service.contract.impl.exec.processors.CustomContractCreationProcessor.start(CustomContractCreationProcessor.java:71)
	at org.hyperledger.besu.evm.processor.AbstractMessageProcessor.process(AbstractMessageProcessor.java:196)
	at com.hedera.node.app.service.contract.impl.exec.FrameRunner.runToCompletion(FrameRunner.java:139)
	at com.hedera.node.app.service.contract.impl.exec.FrameRunner.runToCompletion(FrameRunner.java:90)
	at com.hedera.node.app.service.contract.impl.exec.TransactionProcessor.processTransactionWithParties(TransactionProcessor.java:132)
	at com.hedera.node.app.service.contract.impl.exec.TransactionProcessor.processTransaction(TransactionProcessor.java:108)
	at com.hedera.node.app.service.contract.impl.exec.ContextTransactionProcessor.call(ContextTransactionProcessor.java:127)
	at com.hedera.node.app.service.contract.impl.handlers.ContractCreateHandler.handle(ContractCreateHandler.java:59)
	at com.hedera.node.app.workflows.dispatcher.TransactionDispatcher.dispatchHandle(TransactionDispatcher.java:131)
	at com.hedera.node.app.workflows.handle.DispatchProcessor.tryHandle(DispatchProcessor.java:133)
	at com.hedera.node.app.workflows.handle.DispatchProcessor.processDispatch(DispatchProcessor.java:112)
	at com.hedera.node.app.workflows.standalone.TransactionExecutors.lambda$newExecutor$0(TransactionExecutors.java:226)
	at com.hedera.node.app.workflows.standalone.TransactionExecutors$DefaultTracerBinding.runWhere(TransactionExecutors.java:331)
	at com.hedera.node.app.workflows.standalone.TransactionExecutors.lambda$newExecutor$1(TransactionExecutors.java:225)
	at com.hedera.mirror.web3.service.TransactionExecutionService.execute(TransactionExecutionService.java:80)
	at com.hedera.mirror.web3.service.ContractCallService.doProcessCall(ContractCallService.java:126)
	at com.hedera.mirror.web3.service.ContractCallService.callContract(ContractCallService.java:116)
	at com.hedera.mirror.web3.service.ContractExecutionService.lambda$processCall$0(ContractExecutionService.java:62)
	at com.hedera.mirror.web3.common.ContractCallContext.lambda$run$0(ContractCallContext.java:83)
	at java.base/java.lang.ScopedValue$Carrier$CallableAdapter.call(Unknown Source)
	at java.base/jdk.internal.vm.ScopedValueContainer.callWithoutScope(Unknown Source)
	at java.base/jdk.internal.vm.ScopedValueContainer.call(Unknown Source)
	at java.base/java.lang.ScopedValue$Carrier.get(Unknown Source)
	at java.base/java.lang.ScopedValue.getWhere(Unknown Source)
	at com.hedera.mirror.web3.common.ContractCallContext.run(ContractCallContext.java:83)
	at com.hedera.mirror.web3.service.ContractExecutionService.processCall(ContractExecutionService.java:51)
	at com.hedera.mirror.web3.controller.ContractController.call(ContractController.java:66)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:258)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:191)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:986)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:891)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:914)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:590)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at com.hedera.mirror.web3.config.MetricsFilter.doFilterInternal(MetricsFilter.java:54)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at com.hedera.mirror.web3.config.LoggingFilter.doFilterInternal(LoggingFilter.java:43)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.ServerHttpObservationFilter.doFilterInternal(ServerHttpObservationFilter.java:114)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.ForwardedHeaderFilter.doFilterInternal(ForwardedHeaderFilter.java:173)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:116)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:398)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:903)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1740)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at java.base/java.lang.VirtualThread.run(Unknown Source)`

### Steps to reproduce

Check the Grafana logs on testnet

### Additional context

_No response_

### Hedera network

_No response_

### Version

v0.129.0

### Operating system

None

### PR DESCRIPTION
**Pull request by kselveliev**
**Description**:

Adds per-thread caching using `ContractCallContext` and a configurable buffer to`EntityIdSingleton.get()`
This prevents ID conflicts during entity creation by:

Ensuring consistent IDs within the same flow.
Skipping ahead of recently persisted IDs to avoid collisions.

`EntityIdSingleton.get()` is called twice during single creation flow: once for pending creation and again for validation. If another entity is persisted between these calls, it can cause mismatched IDs and errors like in the ticket log.
We also need a buffer so that cached number does not collide with just recently persisted entity by importer.

* `ContractCallContext` - keeps the cached EntityNumber
* `MirrorNodeEvmProperties` - has a a configurable entityIdReservationHeadroom
* `EntityIdSingleton` - now caches the EntityNumber in context with a buffer applied
* New integration test deploying multiple contracts and current tests modifications

**Related issue(s)**:

Fixes #11166

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by filev94 — DISMISSED** [review]

File: web3/src/test/java/org/hiero/mirror/web3/state/singleton/EntityIdSingletonTest.java [hunk 1]
```diff
@@ -48,7 +54,12 @@ void shouldReturnFirstUserEntityIdWhenMaxIdIsLessThanLastSystemAccount(long shar
         when(commonProperties.getShard()).thenReturn(shard);
         when(commonProperties.getRealm()).thenReturn(realm);
         when(entityRepository.findMaxId(shard, realm)).thenReturn(900L);
-        assertThat(entityIdSingleton.get().number()).isEqualTo(FIRST_USER_ENTITY_ID);
+        try (MockedStatic<ContractCallContext> mockedContext = Mockito.mockStatic(ContractCallContext.class)) {
```
Comment (steven-sheehy) [inline web3/src/test/java/org/hiero/mirror/web3/state/singleton/EntityIdSingletonTest.java:57]: We shouldn't mock simple classes like properties or ContractCallContext. Especially since it causes so much boilerplate with statics. Add `ContextExtension` to `@ExtendWith` and switch to non-mock use of context.

File: web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java [hunk 2]
```diff
@@ -199,6 +199,9 @@ public class MirrorNodeEvmProperties implements EvmProperties {
     @DecimalMax("1.0")
     private double modularizedTrafficPercent = 0.0;
 
+    @Getter
+    private long entityIdReservationHeadroom = 10L;
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:203]: Can you explain why we need this and how you came up with the value of 10? What happens if we just return a cached EntityNumber without adding 10?

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceDeployerTest.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import org.hiero.mirror.web3.web3j.generated.TestContractDeployer;
+import org.junit.jupiter.api.Test;
+
+class ContractCallServiceDeployerTest extends AbstractContractCallServiceTest {
+
+    @Test
+    void testDeployMultipleContracts() throws Exception {
+        final var contract = testWeb3jService.deploy(TestContractDeployer::deploy);
+        final var result = contract.send_deployTestContracts().send();
+        assertThat(result).isNotNull();
```
Comment (steven-sheehy) [inline web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceDeployerTest.java:16]: Is there a way to return the contractID/long zero address to verify the actual number?

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java (hunk 2 above)
Comment (kselveliev) [reply web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:203]: In ticket log, we observed that the first query returned pending creation ID 5992992,
 but a second query shortly after returned 5992993. This likely indicates that a new record was inserted in the database between the two calls.

Currently, if we cache 5992992, validation passes.
However, if another record is inserted using that ID (which was assumed to be the next free one), 
we risk ID collisions. This is problematic because RecordFinalizerBase in hedera.app validates net HBAR balance changes of modified accounts in state at the end of the calls.

Logic in hedera app: 
  ```
      var netHbarBalance = 0L;
        for (final AccountID modifiedAcctId : writableAccountStore.modifiedAccountsInState()) {
            final var modifiedAcct = requireNonNull(writableAccountStore.getAccountById(modifiedAcctId));
            final var persistedAcct = writableAccountStore.getOriginalValue(modifiedAcctId);
            // It's possible the modified account was created in this transaction, in which case the non-existent
            // persisted account effectively has no balance (i.e. its prior balance is 0)
            final var persistedBalance = persistedAcct != null ? persistedAcct.tinybarBalance() : 0;

            // Never allow an account's net hbar balance to be negative
            validateTrue(modifiedAcct.tinybarBalance() >= 0, FAIL_INVALID);
            validateTrue(modifiedAcct.tinybarBalance() <= maxLegalBalance, FAIL_INVALID);

            final var netHbarChange = modifiedAcct.tinybarBalance() - persistedBalance;
            if (netHbarChange != 0) {
                netHbarBalance = addExactOrThrowReason(netHbarBalance, netHbarChange, FAIL_INVALID);
                hbarChanges.put(modifiedAcctId, netHbarChange);
            }
        }
        // Net HBAR balances can only change when creating system entities at genesis
        if (netHbarBalance != 0 && systemEntitiesCreated()) {
            throw new HandleException(FAIL_INVALID);
        }

```

If two accounts end up with the same ID (due to the race condition), the finalizer logic may miscalculate hbar balance, leading to incorrect balance validations or even transaction failures with FAIL_INVALID.

That is the reason why I introduced the headroom config — to ensure we skip a buffer of IDs and avoid potential clashes with concurrently created accounts.

Those newly created accounts with headroom in id will only be in memory so they will not cause a problem if they have slightly larger ids since they wont be persisted.

Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:203]: If we analyze execution on transactions in consensus nodes - we have to pay roughly 56K gas for a hollow account creation and we have a limit of 15M gas per block. This means on theory we can have roughly 300 new IDs created in DB for a single block. If an `estimateGas` simulation takes 2s or more (which should be nearly impossible) we can have a collision with up to 300 newly created IDs.

If we use this buffer approach, since we won't save this value in DB but use it only for caching new entity IDs what about increasing the buffer even more to cover negative corner cases. E.g. the buffer can be 1000. What do you think @steven-sheehy ?

File: web3/src/test/java/org/hiero/mirror/web3/state/singleton/EntityIdSingletonTest.java (hunk 1 above)
Comment (kselveliev) [reply web3/src/test/java/org/hiero/mirror/web3/state/singleton/EntityIdSingletonTest.java]: Addressed

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceDeployerTest.java (hunk 3 above)
Comment (kselveliev) [reply web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceDeployerTest.java:16]: Not sure that we can do that, we can modify the contract to return the addresses of the deployed contracts. But they will be the canonical EVM addresses. I don't see a way to derive the long zero address from the canonical EVM since they are in memory.

File: web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java (hunk 2 above)
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:203]: Need to document in configuration.md

Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:203]: Maybe a simpler name like
```suggestion
    private long entityNumBuffer = 1000L;
```

Comment (kselveliev) [reply web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java]: Addressed

Comment (kselveliev) [reply web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java]: Addressed

Comment (kselveliev) [reply web3/src/main/java/org/hiero/mirror/web3/evm/properties/MirrorNodeEvmProperties.java]: Addressed

**Review by filev94 — APPROVED** [review]

**Review by steven-sheehy — APPROVED** [review]
